expr/field_reference.go

```diff
@@ -401,6 +401,34 @@ func (f *FieldReference) Visit(VisitFunc) Expression {
 }
 func (*FieldReference) IsScalar() bool { return true }
 
+// argument is only utilized if root is not an expression
+func (f *FieldReference) UpdateType(baseSchema types.Type) error {
+	switch rt := f.Reference.(type) {
+	case ReferenceSegment:
+		var rootType types.Type
+
+		if f.Root == RootReference {
+			if baseSchema == nil {
+				return fmt.Errorf("%w: updating type for root reference requires base schema",
+					substraitgo.ErrInvalidType)
+			}
+			rootType = baseSchema
+		} else if rootExpr, ok := f.Root.(Expression); ok {
+			rootType = rootExpr.GetType()
+		}
+
+		typ, err := rt.GetType(rootType)
+		if err != nil {
+			return err
+		}
+		f.knownType = typ
+		return nil
+	case *MaskExpression:
```

Comment:
Not a big deal, but any reason for this case *MaskExpression

Reply:
it'll end up falling through to the `NotImplemented` error for now. I just wanted to make it obvious that the `MaskExpression` case wasn't implemented yet here so it doesn't get lost or forgotten about :smile: